R/sql-query.R

```diff
@@ -103,7 +103,22 @@ join_vars <- function(x_names, y_names, by, suffix = c(".x", ".y")) {
   y_new <- y_names
   y_new[y_match] <- paste0(y_names[y_match], suffix$y)
 
-  list(x = setNames(x_new, x_names), y = setNames(y_new, y_names))
+  # In left and inner joins, return key values only from x
+  # In right joins, return key values only from y
+  # In full joins, return key values by coalescing values from x and y
+  x_x <- x_names
+  x_y <- by$y[match(x_names, by$x)]
+  x_y[type == "left" | type == "inner"] <- NA
+  x_x[type == "right" & !is.na(x_y)] <- NA
+  y_x <- rep_len(NA, length(y_names))
+  y_y <- y_names
```

Comment:
Line by line explanation:

109. `x_x` is a vector of names of columns in the left table corresponding to the column names in `x_names`.
110. `x_y` is a vector of names of columns in the right table corresponding to the column names in `x_names`.
111. `x_y` is all `NA` for left and inner joins, because values from columns in the left table should come only from the left table.
112. For right joins, the positions of `x_x` representing key columns are `NA`, because values in key columns should come from the right table. (For full joins, the positions of `x_x` and `x_y` representing key columns are both non-`NA`, because we want to coalesce values from the left and right tables.)
113. `y_x` is a vector of names of columns in the left table corresponding to the column names in `y_names`. It's always all `NA`.
114. `y_y` is a vector of names of columns in the right table corresponding to the column names in `y_names`. It's always all non-`NA` because columns in the right table always come from the right table.